File: sechat/ios/Classes/channels/FileTransferChannel.m
//
//  FileTransferChannel.m
//  Sechat
//
//  Created by Albert Moky on 2023/5/7.
//

#import <DIMClient/DIMClient.h>

#import "FileTransfer.h"
#import "ChannelManager.h"

#import "FileTransferChannel.h"

static void onMethodCall(FlutterMethodCall* call, FlutterResult result);

static NSMutableDictionary *uploadInfo(DIMUploadRequest *request) {
    NSDictionary *info;
    if ([request isKindOfClass:[DIMUploadTask class]]) {
        DIMUploadTask *task = (DIMUploadTask *)request;
        info = @{
            @"api": [task.url absoluteString],
            @"name": task.name,
            @"filename": task.filename,
        };
    } else {
        info = @{
            @"api": [request.url absoluteString],
            @"name": request.name,
            @"path": request.path,
            @"sender": [request.sender string],
        };
    }
    return [info mutableCopy];
}

@implementation FileTransferChannel

+ (instancetype)channelWithName:(NSString*)name
                binaryMessenger:(NSObject<FlutterBinaryMessenger>*)messenger
                          codec:(NSObject<FlutterMethodCodec>*)codec {
    id channel = [[self alloc] initWithName:name
                            binaryMessenger:messenger
                                      codec:codec];
    [channel setMethodCallHandler:^(FlutterMethodCall *call, FlutterResult result) {
        onMethodCall(call, result);
    }];
    return channel;
}

#pragma mark DIMUploadDelegate

- (void)uploadTask:(__kindof DIMUploadRequest *)req onSuccess:(NSURL *)url {
    NSLog(@"onUploadSuccess: %@, url: %@", req, url);
    NSMutableDictionary *params = uploadInfo(req);
    [params setObject:@{
        @"url": [url absoluteString],
    } forKey:@"response"];
    [self invokeMethod:FtpChannelOnUploadSuccess arguments:params];
}

- (void)uploadTask:(__kindof DIMUploadRequest *)req onFailed:(NSException *)error {
    NSLog(@"onUploadFailed: %@, error: %@", req, error);
    NSMutableDictionary *params = uploadInfo(req);
    [params setObject:[error description] forKey:@"error"];
    [self invokeMethod:FtpChannelOnUploadFailure arguments:params];
}

- (void)uploadTask:(__kindof DIMUploadRequest *)req onError:(NSError *)error {
    NSLog(@"onUploadError: %@, error: %@", req, error);
    NSMutableDictionary *params = uploadInfo(req);
    [params setObject:[error description] forKey:@"error"];
    [self invokeMethod:FtpChannelOnUploadFailure arguments:params];
}

#pragma mark DIMDownloadDelegate

- (void)downloadTask:(__kindof DIMDownloadRequest *)req onSuccess:(NSString *)path {
    NSLog(@"onDownloadSuccess: %@, path: %@", req, path);
    NSDictionary *params = @{
        @"url": [req.url absoluteString],
        @"path": path,
    };
    [self invokeMethod:FtpChannelOnDownloadSuccess arguments:params];
}

- (void)downloadTask:(__kindof DIMDownloadRequest *)req onFailed:(NSException *)error {
    NSLog(@"onDownloadFailed: %@, error: %@", req, error);
    NSDictionary *params = @{
        @"url": [req.url absoluteString],
        @"path": req.path,
        @"error": [error description],
    };
    [self invokeMethod:FtpChannelOnDownloadFailure arguments:params];
}

- (void)downloadTask:(__kindof DIMDownloadRequest *)req onError:(NSError *)error {
    NSLog(@"onDownloadFailed: %@, error: %@", req, error);
    NSDictionary *params = @{
        @"url": [req.url absoluteString],
        @"path": req.path,
        @"error": [error description],
    };
    [self invokeMethod:FtpChannelOnDownloadFailure arguments:params];
}


@end

static inline void onMethodCall(FlutterMethodCall* call, FlutterResult success) {
    FileTransfer *ftp = [FileTransfer sharedInstance];
    NSString *method = [call method];
    if ([method isEqualToString:FtpChannelDownloadFile]) {
        // downloadEncryptedFile
        NSURL *url = [NSURL URLWithString:[call.arguments objectForKey:@"url"]];
        NSString *path = [ftp downloadEncryptedData:url];
        success(path);
    } else if ([method isEqualToString:FtpChannelDownloadAvatar]) {
        // downloadAvatar
        NSURL *url = [NSURL URLWithString:[call.arguments objectForKey:@"url"]];
        NSString *path = [ftp downloadAvatar:url];
        success(path);
    } else if ([method isEqualToString:FtpChannelUploadFile]) {
        // uploadEncryptFile
        FlutterStandardTypedData *data = [call.arguments objectForKey:@"data"];
        NSString *filename = [call.arguments objectForKey:@"filename"];
        id<MKMID> sender = MKMIDParse([call.arguments objectForKey:@"sender"]);
        NSURL *url = [ftp uploadEncryptedData:[data data] filename:filename sender:sender];
        success([url absoluteString]);
    } else if ([method isEqualToString:FtpChannelUploadAvatar]) {
        // uploadAvatar
        FlutterStandardTypedData *data = [call.arguments objectForKey:@"data"];
        NSString *filename = [call.arguments objectForKey:@"filename"];
        id<MKMID> sender = MKMIDParse([call.arguments objectForKey:@"sender"]);
        NSURL *url = [ftp uploadAvatar:[data data] filename:filename sender:sender];
        success([url absoluteString]);
    } else if ([method isEqualToString:FtpChannelSetUploadAPI]) {
        // setUploadAPI
        NSString *api = [call.arguments objectForKey:@"api"];
        NSString *secret = [call.arguments objectForKey:@"secret"];
        if (api) {
            ftp.api = api;
        }
        if (secret) {
            ftp.secret = secret;
        }
        success(nil);
    } else if ([method isEqualToString:FtpChannelGetCachesDirectory]) {
        // getCachesDirectory
        NSString *dir = [DIMStorage cachesDirectory];
        success(dir);
    } else if ([method isEqualToString:FtpChannelGetTemporaryDirectory]) {
        // getCachesDirectory
        NSString *dir = [DIMStorage temporaryDirectory];
        success(dir);
    } else {
        NSLog(@"not implemented: %@", method);
        assert(false);
    }
}

File: sechat/android/app/src/main/java/chat/dim/channels/FileTransferChannel.java
package chat.dim.channels;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import java.io.IOError;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodCodec;

import chat.dim.filesys.LocalCache;
import chat.dim.http.FileTransfer;
import chat.dim.http.DownloadDelegate;
import chat.dim.http.DownloadRequest;
import chat.dim.http.UploadDelegate;
import chat.dim.http.UploadRequest;
import chat.dim.http.UploadTask;
import chat.dim.protocol.ID;
import chat.dim.utils.Log;

public class FileTransferChannel extends MethodChannel implements UploadDelegate, DownloadDelegate {

    public FileTransferChannel(@NonNull BinaryMessenger messenger, @NonNull String name, @NonNull MethodCodec codec) {
        super(messenger, name, codec);
        setMethodCallHandler(new FileChannelHandler());
    }

    //-------- Upload Delegate

    private static Map<String, Object> uploadInfo(UploadRequest request) {
        Map<String, Object> info = new HashMap<>();
        // request info
        if (request instanceof UploadTask) {
            UploadTask task = (UploadTask) request;
            info.put("api", request.url.toString());
            info.put("name", request.name);
            info.put("filename", task.filename);
        } else {
            info.put("api", request.url.toString());
            info.put("path", request.path);
            info.put("name", request.name);
            info.put("sender", request.sender.toString());
        }
        return info;
    }

    @Override
    public void onUploadSuccess(UploadRequest request, URL url) {
        Log.info("onUploadSuccess: " + request + ", url: " + url);
        Map<String, Object> response = new HashMap<>();
        response.put("url", url.toString());
        Map<String, Object> params = uploadInfo(request);
        params.put("response", response);
        new Handler(Looper.getMainLooper()).post(() ->
                invokeMethod(ChannelMethods.ON_UPLOAD_SUCCESS, params));
    }

    @Override
    public void onUploadFailed(UploadRequest request, IOException error) {
        Log.error("onUploadFailed: " + request + ", error: " + error);
        Map<String, Object> params = uploadInfo(request);
        params.put("error", error.toString());
        new Handler(Looper.getMainLooper()).post(() ->
                invokeMethod(ChannelMethods.ON_UPLOAD_FAILURE, params));
    }

    @Override
    public void onUploadError(UploadRequest request, IOError error) {
        Log.error("onUploadError: " + request + ", error: " + error);
        Map<String, Object> params = uploadInfo(request);
        params.put("error", error.toString());
        new Handler(Looper.getMainLooper()).post(() ->
                invokeMethod(ChannelMethods.ON_UPLOAD_FAILURE, params));
    }

    //-------- Download Delegate

    @Override
    public void onDownloadSuccess(DownloadRequest request, String path) {
        Log.info("onDownloadSuccess: " + request + ", path: " + path);
        Map<String, Object> params = new HashMap<>();
        params.put("url", request.url.toString());
        params.put("path", request.path);
        new Handler(Looper.getMainLooper()).post(() ->
                invokeMethod(ChannelMethods.ON_DOWNLOAD_SUCCESS, params));
    }

    @Override
    public void onDownloadFailed(DownloadRequest request, IOException error) {
        Log.info("onDownloadFailed: " + request + ", error: " + error);
        Map<String, Object> params = new HashMap<>();
        params.put("url", request.url.toString());
        params.put("path", request.path);
        params.put("error", error.toString());
        new Handler(Looper.getMainLooper()).post(() ->
                invokeMethod(ChannelMethods.ON_DOWNLOAD_FAILURE, params));
    }

    @Override
    public void onDownloadError(DownloadRequest request, IOError error) {
        Log.info("onDownloadFailed: " + request + ", error: " + error);
        Map<String, Object> params = new HashMap<>();
        params.put("url", request.url.toString());
        params.put("path", request.path);
        params.put("error", error.toString());
        new Handler(Looper.getMainLooper()).post(() ->
                invokeMethod(ChannelMethods.ON_DOWNLOAD_FAILURE, params));
    }

    static class FileChannelHandler implements MethodChannel.MethodCallHandler {

        @Override
        public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
            FileTransfer ftp = FileTransfer.getInstance();
            switch (call.method) {
                case ChannelMethods.DOWNLOAD_FILE: {
                    URL url = getURL(call);
                    String path = ftp.downloadEncryptedFile(url);
                    result.success(path);
                    break;
                }
                case ChannelMethods.DOWNLOAD_AVATAR: {
                    URL url = getURL(call);
                    String path = ftp.downloadAvatar(url);
                    result.success(path);
                    break;
                }
                case ChannelMethods.UPLOAD_FILE: {
                    byte[] data = call.argument("data");
                    String filename = call.argument("filename");
                    ID sender = ID.parse(call.argument("sender"));
                    URL url = ftp.uploadEncryptData(data, filename, sender);
                    result.success(url == null ? null : url.toString());
                    break;
                }
                case ChannelMethods.UPLOAD_AVATAR: {
                    byte[] data = call.argument("data");
                    String filename = call.argument("filename");
                    ID sender = ID.parse(call.argument("sender"));
                    URL url = ftp.uploadAvatar(data, filename, sender);
                    result.success(url == null ? null : url.toString());
                    break;
                }
                case ChannelMethods.SET_UPLOAD_API: {
                    ftp.api = call.argument("api");
                    ftp.secret = call.argument("secret");
                    result.success(null);
                    break;
                }
                case ChannelMethods.GET_CACHES_DIRECTORY: {
                    LocalCache localCache = LocalCache.getInstance();
                    String dir = localCache.getCachesDirectory();
                    result.success(dir);
                    break;
                }
                case ChannelMethods.GET_TEMPORARY_DIRECTORY: {
                    LocalCache localCache = LocalCache.getInstance();
                    String dir = localCache.getTemporaryDirectory();
                    result.success(dir);
                    break;
                }
                default:
                    result.notImplemented();
                    break;
            }
        }

        private URL getURL(MethodCall call) {
            String urlString = call.argument("url");
            try {
                return new URL(urlString);
            } catch (MalformedURLException e) {
                Log.error("URL error: " + e);
                return null;
            }
        }

    }
}
